pkgs/development/python-modules/websockets/default.nix

```diff
@@ -23,11 +23,25 @@ buildPythonPackage rec {
   # Tests fail on Darwin with `OSError: AF_UNIX path too long`
   doCheck = !stdenv.isDarwin;
 
-  # Disable all tests that need to terminate within a predetermined amount of
-  # time. This is nondeterministic.
   patchPhase = ''
+    # Disable all tests that need to terminate within a predetermined amount of
+    # time. This is nondeterministic.
     sed -i 's/with self.assertCompletesWithin.*:/if True:/' \
       tests/legacy/test_protocol.py
+
+    # Disables tests relying on tight timeouts to avoid failures like:
+    #   File "/build/source/tests/legacy/test_protocol.py", line 1270, in test_keepalive_ping_with_no_ping_timeout
+    #     ping_1_again, ping_2 = tuple(self.protocol.pings)
+    #   ValueError: too many values to unpack (expected 2)
+    for t in \
+             test_keepalive_ping_stops_when_connection_closing \
+             test_keepalive_ping_does_not_crash_when_connection_lost \
+             test_keepalive_ping \
+             test_keepalive_ping_not_acknowledged_closes_connection \
+             test_keepalive_ping_with_no_ping_timeout \
+      ; do
+      sed -i "s/def $t(/def skip_$t(/" tests/legacy/test_protocol.py
+    done
```

Comment:
since this uses unittest suites, we could just use the pytestCheckHook, and use its `disableTests`

Reply:
I'll need some help with that. Do I need to pass `discover` argument to `-m pytest`? Here is my attempt:

```diff
--- a/pkgs/development/python-modules/websockets/default.nix
+++ b/pkgs/development/python-modules/websockets/default.nix
@@ -6,2 +6,3 @@
 , pythonOlder
+, pytestCheckHook
 }:
@@ -46,7 +47,3 @@ buildPythonPackage rec {

-  checkPhase = ''
-    runHook preCheck
-    ${python.interpreter} -m unittest discover
-    runHook postCheck
-  '';
+  checkInputs = [ pytestCheckHook ];
```

Looks like it breaks a testing environment:

```
$ nix build -f. python3Packages.websockets -L
...
python3.9-websockets> ==================================== ERRORS ====================================
python3.9-websockets> __________________ ERROR collecting compliance/test_client.py __________________
python3.9-websockets> compliance/test_client.py:48: in <module>
python3.9-websockets>     asyncio.run(run_tests(SERVER, urllib.parse.quote(AGENT)))
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/runners.py:44: in run
python3.9-websockets>     return loop.run_until_complete(main)
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/base_events.py:642: in run_until_complete
python3.9-websockets>     return future.result()
python3.9-websockets> compliance/test_client.py:40: in run_tests
python3.9-websockets>     cases = await get_case_count(server)
python3.9-websockets> compliance/test_client.py:21: in get_case_count
python3.9-websockets>     async with websockets.connect(uri) as ws:
python3.9-websockets> /nix/store/8n7xjx84q7v5iqqrd87nz4ymdb0m4m04-python3.9-websockets-10.1/lib/python3.9/site-packages/websockets/legacy/client.py:633: in __aenter__
python3.9-websockets>     return await self
python3.9-websockets> /nix/store/8n7xjx84q7v5iqqrd87nz4ymdb0m4m04-python3.9-websockets-10.1/lib/python3.9/site-packages/websockets/legacy/client.py:650: in __await_impl_timeout__
python3.9-websockets>     return await asyncio.wait_for(self.__await_impl__(), self.open_timeout)
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/tasks.py:481: in wait_for
python3.9-websockets>     return fut.result()
python3.9-websockets> /nix/store/8n7xjx84q7v5iqqrd87nz4ymdb0m4m04-python3.9-websockets-10.1/lib/python3.9/site-packages/websockets/legacy/client.py:654: in __await_impl__
python3.9-websockets>     transport, protocol = await self._create_connection()
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/base_events.py:1056: in create_connection
python3.9-websockets>     raise exceptions[0]
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/base_events.py:1041: in create_connection
python3.9-websockets>     sock = await self._connect_sock(
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/base_events.py:955: in _connect_sock
python3.9-websockets>     await self.sock_connect(sock, address)
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/selector_events.py:502: in sock_connect
python3.9-websockets>     return await fut
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/selector_events.py:537: in _sock_connect_cb
python3.9-websockets>     raise OSError(err, f'Connect call failed {address}')
python3.9-websockets> E   ConnectionRefusedError: [Errno 111] Connect call failed ('127.0.0.1', 8642)
python3.9-websockets> __________________ ERROR collecting compliance/test_server.py __________________
python3.9-websockets> compliance/test_server.py:29: in <module>
python3.9-websockets>     asyncio.run(main())
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/runners.py:44: in run
python3.9-websockets>     return loop.run_until_complete(main)
python3.9-websockets> /nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/lib/python3.9/asyncio/base_events.py:642: in run_until_complete
python3.9-websockets>     return future.result()
python3.9-websockets> compliance/test_server.py:22: in main
python3.9-websockets>     with websockets.serve(echo, HOST, PORT, max_size=2 ** 25, max_queue=1):
python3.9-websockets> E   AttributeError: __enter__
python3.9-websockets> =========================== short test summary info ============================
python3.9-websockets> ERROR compliance/test_client.py - ConnectionRefusedError: [Errno 111] Connect...
python3.9-websockets> ERROR compliance/test_server.py - AttributeError: __enter__
python3.9-websockets> !!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!
```

Reply:
hmm, could be an issue. could probably visit this later